cpp/src/cylon/net/mpi/mpi_operations.hpp

```diff
@@ -31,6 +35,54 @@ cylon::Status AllReduce(const void *send_buf,
                         const std::shared_ptr<DataType> &data_type,
                         cylon::net::ReduceOp reduce_op);
 
+/**
+ * check whether the invoking worker is the root
+ * @param root
+ * @param ctx
+ * @return
+ */
+bool AmIRoot(int root, std::shared_ptr<cylon::CylonContext> ctx);
```

Comment:
i think we should send CylonContext to methods as: 
    std::shared_ptr<const cylon::CylonContext> ctx
this guarantees that CylonContex will not be modified in methods. 
For this though, CylonContex methods such as GetRank and GetWorldSize has to be const. 
Not sure abut sending ctx as an address.

Reply:
when we have `std::shared_ptr<cylon::CylonContext> ctx`, the shared_ptr will be copied into the method (pass by copy). So, what I am suggesting is, let's pass it by reference (and this case, const reference). Because there's no reason for us to make a copy of the shared_ptr. We just need to access its methods. 

I dont think the 'constness' of `GetRank` and `GetWorldSize` matters here because this method is not a class method. But I agree, that the should be const methods in `CylonContext`

Reply:
they recommend sending shared_ptr as const reff as you suggest: 
https://stackoverflow.com/questions/3310737/should-we-pass-a-shared-ptr-by-reference-or-by-value

maybe we can have a typedef for CylonContext  as:
using CylonCtxPtr = std::shared_ptr<const CylonContext>;